avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java

```diff
@@ -544,11 +569,24 @@ private static Object maybeWrapSchemaless(Schema schema, Object value, String ty
   }
 
   public org.apache.avro.Schema fromConnectSchema(Schema schema) {
+    return fromConnectSchema(schema, new HashMap<String, org.apache.avro.Schema>());
+  }
+
+  public org.apache.avro.Schema fromConnectSchema(Schema schema, Map<String, org.apache.avro.Schema> schemaMap) {
+    return fromConnectSchema(schema, schemaMap, false);
+  }
+
+
+  public org.apache.avro.Schema fromConnectSchema(Schema schema, Map<String, org.apache.avro.Schema> schemaMap, boolean ignoreOptional) {
     if (schema == null) {
       return ANYTHING_SCHEMA;
     }
 
     org.apache.avro.Schema cached = fromConnectSchemaCache.get(schema);
+
+    if (cached == null && !AVRO_TYPE_UNION.equals(schema.name()) && !schema.isOptional()) {
```

Comment:
So there is two things going on here, caching the converted schema's so that when we get the same connect schema we don't have to re-convert its a simple cache lookup (original case). 

What we added was as you've noted a map of internal schemas to avoid type re-declaration if a type is reused this actually blows up if you don't do this and have a type used in multiple places. 

Also it only holds reference the non-optional as technically an optional is actually a union of null and the non-opitonal, which if used in multiple places some optional some non-optional will cause error as you redefine type.